keps/sig-node/3673-kubelet-parallel-image-pull-limit/README.md

```diff
@@ -0,0 +1,862 @@
+* The default value of `maxParallelImagePulls` will be set to 0, which applies no limit on parallel pulls, the same as current behavior.
+* When `serialize-image-pulls` is true, `maxParallelImagePulls` will be ignored. A warning message will be logged by kubelet if `serialize-image-pulls` is true and `maxParallelImagePulls` is larger than 1.
+
+### User Stories (Optional)
+
+<!--
+Detail the things that people will be able to do if this KEP is implemented.
+Include as much detail as possible so that people can understand the "how" of
+the system. The goal here is to make this feel real for users without getting
+bogged down.
+-->
+
+#### Story 1
+A kubernetes user wants to enable parallel image pulls to reduce workload startup latency, but they don't want too many images to be pulled at the same time, as that might burden the disk IO and therefore has a negative impact on disk performance, or might use too much network bandwidth. By setting a proper `maxParallelImagePulls`, they will have a better control over the parallelism and therefore avoid overwhelming the disk.
+
+
+### Notes/Constraints/Caveats (Optional)
+
+<!--
+What are the caveats to the proposal?
+What are some important details that didn't come across above?
+Go in to as much detail as necessary here.
+This might be a good place to talk about core concepts and how they relate.
+-->
+
+### Risks and Mitigations
+
+<!--
+What are the risks of this proposal, and how do we mitigate? Think broadly.
+For example, consider both security and how this will impact the larger
+Kubernetes ecosystem.
+
+How will security be reviewed, and by whom?
+
+How will UX be reviewed, and by whom?
+
+Consider including folks who also work outside the SIG or subproject.
+-->
+
+Since this feature is purely in-memory, if kubelet restarts, it will lose track of the image pulls before the restart and therefore allow the next N (N is the value of `maxParallelImagePulls`) image pulls to proceed.
+In worst case, there could be 2*N image pulls in parallel, if kubelet restarts in the middle.
+
+## Design Details
+
+<!--
+This section should contain enough information that the specifics of your
+change are understandable. This may include API specs (though not always
+required) or even code snippets. If there's any ambiguity about HOW your
+proposal will be implemented, this is the place to discuss them.
+-->
+
+* The implementation will be similar to the existing `serialImagePuller`. More specifically:
+  * `maxParallelImagePulls` will be passed to [parallelImagePuller](https://github.com/kubernetes/kubernetes/blob/054ceab58d803df0c4910fa722138f0ed2d94701/pkg/kubelet/images/puller.go#L41), and `parallelImagePuller` will create a channel of the size of `maxParallelImagePulls`.
+  * `parallelImagePuller.pullImage` will simply try to send the pull request to the channel, and will be blocked if the channel is full.
+  * A go routine will keep processing the requests in the channel.
```

Comment:
Is it possible for the same image pull to be requested multiple times? I.e. can you have a queue of pull requests like: A, B, C, D, A?  Is there anything needed to detect that when the 1st pull for A completes that other pull requests in the queue are satisfied?

Reply:
Good point, it is possible technically.  The second A will be enqueued if the first A hasn't completed. But I don't think we really need extra mechanism to prevent the second A from being queued, because 
1. This is the same behavior as the current serial image pulling, i.e. the second A will be queued and waiting if the first A is in progress.
2. When the second A is processed by the puller, image service will return immediately because the image is pulled. So overall this will not cause too much latency.

Reply:
+1 - I was thinking about that too, but I don't see a problem with it